pkg/controller/sparkapplication/controller.go

```diff
@@ -1001,3 +1009,16 @@ func (c *Controller) hasApplicationExpired(app *v1beta2.SparkApplication) bool {
 
 	return false
 }
+
+// Clean up when the spark application is terminated.
+func (c *Controller) cleanUpOnTermination(oldApp, newApp *v1beta2.SparkApplication) error {
+	if needScheduling, scheduler := c.shouldDoBatchScheduling(newApp); needScheduling {
+		// batch schduler is cleaned up only when state is changed to completion state
+		if newApp.Status.AppState.State != oldApp.Status.AppState.State {
```

Comment:
This check is not needed if deletion is idempotent, i.e., if we ignore NotFound error above.

Reply:
looks reasonable
updated to
```go
if needScheduling, scheduler := c.shouldDoBatchScheduling(newApp); needScheduling {
	if err := scheduler.CleanupOnCompletion(newApp); err != nil && !errors.IsNotFound(err) {
		return err
	}
}
```